### ISSUE TITLE
Can't hide modal before it is "shown?"

### ISSUE THREAD

**Issue by binseq** [twbs/bootstrap#34213]
In Bootstrap 5, calling Modal.hide() has no effect until modal is "shown" (it works as expected if the "fade" css class is removed). 

Suppose you have list of items and a single modal.When you click on each item, details about that item are shown in the modal, but you accidentally clicked the wrong item, so you want to quickly cancel by clicking the backdrop while it's fading in. Currently you cannot do that. You have to wait until the wrong item is shown to you before you can close it. 

Is there a way around this?

Maybe there should be an optional parameter, Modal.hide(force=true).

I've made a simple demo, you can just copy/paste and run it...
```
<!doctype html>
<html lang="en">
<head>
	<meta charset="utf-8">
	<title>Modal Test</title>
	<meta name="viewport" content="width=device-width, initial-scale=1">
	<link href="https://cdn.jsdelivr.net/npm/bootstrap@5.0.1/dist/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-+0n0xVW2eSR5OomGNYDnhzAbDsOXxcvSN1TPprVMTNDbiYZCxYbOOl7+AMvyTG2x" crossorigin="anonymous">

	<style>
		.show {
			 transition: opacity 500ms;
		}
	</style>

</head>
<body>

	<p>
		Click a button and quickly try to cancel the modal by clicking the backdrop.
	</p>

	<!-- Buttons -->
	<button type="button" class="btn btn-primary item-button" data-item-info="You clicked # 1">
		Item # 1
	</button>

	<button type="button" class="btn btn-primary item-button" data-item-info="You clicked # 2">
		Item # 2
	</button>

	<button type="button" class="btn btn-primary item-button" data-item-info="You clicked # 3">
		Item # 3
	</button>

	<!-- Modal -->
	<div class="modal fade" id="myModal" tabindex="-1" aria-labelledby="myModalLabel" aria-hidden="true">
		<div class="modal-dialog">
			<div class="modal-content">
				<div class="modal-body">
					...
				</div>
			</div>
		</div>
	</div>

	<script src="https://cdn.jsdelivr.net/npm/bootstrap@5.0.1/dist/js/bootstrap.bundle.min.js" integrity="sha384-gtEjrD/SeCtmISkJkNUaaKMoLD0//ElJ19smozuHV6z3Iehds+3Ulb9Bn9Plx0x4" crossorigin="anonymous"></script>
	<script src="https://ajax.googleapis.com/ajax/libs/jquery/3.6.0/jquery.min.js"></script>
	<script>
		var myModal = new bootstrap.Modal(document.getElementById('myModal'), {
			keyboard: false
		})

		$('.item-button').click(function(){
			$('#myModal').find('.modal-body').text($(this).data('item-info'))
			myModal.show();
		})

		$('.modal').click(function(e){
			if(!$(e.target).hasClass("modal-body")){
				console.log('Clicked modal backdrop, calling .hide()...')
				myModal.hide();
			}
		})

	</script>

</body>
</html>
```

**Comment by mdo** [comment]
Possibly a duplicate of #34055?

**Comment by GeoSot** [comment]
Try to use the following scripts on your test:

```
https://deploy-preview-34085--twbs-bootstrap.netlify.app/docs/5.0/dist/js/bootstrap.bundle.min.js

https://deploy-preview-34085--twbs-bootstrap.netlify.app/docs/5.0/dist/css/bootstrap.min.css
```

I suppose it will work fine. (and please feedback us)

**Comment by binseq** [comment]
The solution to remove the if block: `if (!this._isShown || this._isTransitioning) {`
https://github.com/twbs/bootstrap/blob/b613c8dbb6c30e62adaefc669130549fd9c2e28b/js/src/modal.js#L151

**Comment by GeoSot** [comment]
> 
> 
> The solution is to remove the if block: `if (!this._isShown || this._isTransitioning) {`
> 
> https://github.com/twbs/bootstrap/blob/b613c8dbb6c30e62adaefc669130549fd9c2e28b/js/src/modal.js#L151

or by genlty  initializing `this.isShown = this._element.classList.contains(CLASS_NAME_SHOW)` .

**Comment by alpadev** [comment]
@GeoSot I'm afraid #34085 wouldn't fix that. The problem here is that while the modal is transitioning it just returns from calls to `.show()` and `.hide()` and does nothing.

Simply removing `_isTransitioning` like in #34232 causes the events to fire in wrong order e.g. `shown` is dispatched after the `hide` event. If some 3rd party makes use of them, it may break their implementation. So IMHO this is not the proper way to do it.

**Comment by GeoSot** [comment]
but it is related to #33727 . Right?

**Comment by binseq** [comment]
> Here is the OPs code with his _fix_ from #34232 applied https://jsfiddle.net/z6p0oujt/. IMO the improvement from the that fix is negligible as it requires a certain timing to make use of the change at all. (You have to wait for the backdrop before you click and it saves you some ms that is needed to fully show the modal.)
> 
> If it happens on a frequent basis, one can simply remove the fade class so it wouldn't be a problem at all.

It took me a while to figure out why the demo I posted wasn't working with the updated fix, then I realized a small change to the CSS was required for it to work.
Try it now: https://jsfiddle.net/fgyuerLo/

Also, this was just a simple demo of one use case, least complex example that could think of. The reason I need hide to respond instantly is I want to hide the modal when user navigates back in the browser. I do that by modifying the URL (with pushState) with a special marker to indicate that .show() has been called. For example, pushState PREVIOUS_URL + #modal. Then I listen for popstate event and examine the url. If the #modal hash is missing, then I know the user navigated back so I call .hide(). Now the problem is if user navigates back before hide() is ready to respond, the browser is already on PREVIOUS_URL but modal is still showing. Now if user navigates back again, the browser is back two pages.

**Comment by ddanielou** [comment]
@binseq I faced a similar problem with offcanvas recently. I ended up listening once for the "shown" event before calling `.hide()`, e.g.
```javascript
function forceHide(el){
    el.addEventListener('shown.bs.modal', () => bootstrap.Modal.getInstance(el).hide(), {once: true});
}
```
Though that wouldn't help get rid of the small delay, at least `.hide()` would work. It's also somewhat logical, as you (arguably) wouldn't hide something that isn't shown yet.

**Comment by alpadev** [comment]
> But it is related to #33727 . Right?

Yes, somewhat. The problem is a bit different but it got something to do with transitions being asynchronous and methods are called in a synchronously. We would need to implement some way of cancelling transitions or to automatically queue code that is called during busy states.

**Comment by pauljura** [comment]
Hi team, I want to add my support for a resolution to this issue.

My use case is that I display a table of items, and clicking on at item will display a large modal with an edit screen. The content of the modal gets loaded asynchronously on 'show.bs.modal'. However, if loading the modal content fails (e.g. session has timed out, or permission error) then I want to display a different, smaller modal with an error message. Since I can't display two modals at once, I have to close the first one. But if the error happens quickly, before the main modal has finished transitioning, then I can't close it.

I would very much like a single method to both hide an already open modal and also hide it immediately if it's in the middle of being shown.

You might wonder why I don't just display the error message in the main modal... It's because that one has a custom toolbar and buttons intended for editing, so if I was going to show an error message then I would want to hide these elements, and then reveal them again if the modal gets re-opened. The other workaround is to just manually hide the entire modal. Either way feels hacky.
